working/0698 - Enhanced Default Constructors/proposal.md

```diff
@@ -0,0 +1,166 @@
+# Dart Default Constructor Enhancement
```

Comment:
This looks very cool.  It would be good to have a section discussing the impact on breaking changes.  I'm a little concerned about the fact that this makes small changes to the class or superclass show up invisibly in the constructor in ways that are breaking.  In the superclass case, I think that means that more things are breaking changes.  In the class case, I think (but am not sure) that you can always rewrite the constructor explicitly to avoid the breaking change, but that's easy not to notice.  Maybe that's an advantage of the primary constructor syntax of Kotlin?

Reply:
I really like this feature but I think one important thing is misses is what I call, for the lack of better word, "stability" (in parallel with mathematical notion of [stability](https://en.wikipedia.org/wiki/Stability_theory)) - if you want to make small semantic change, you can achieve it with a small syntactic change. 

For example, imagine I want a constructor that behaves like "enhanced default constructor" but has few `assert`-s. 

I can't do that with few changes - I suddenly need to create the full constructor which initialises all the fields explicitly. 

Kotlin constructor syntax has stability property because features their combine and build on top of each other - rather than being "exclusive" (you either have enhanced constructor - or old school constructor... why not both?)

Reply:
Stability, non-chaotic syntax, ease of migration, etc. It's a good idea.

It's *hard* to do here because we are automatically generating a significant amount of syntax.

Maybe we could use the `default` word in the parameter list:

```dart
  Foo(default) : super.bar;
```
The `default` stands for named initializers for all uninitialized public non-late fields of the current class plus forwarding parameters for all parameters of the `super.bar` constructor (with some conflict handling).

Reply:
> Maybe we could use the `default` word in the parameter list:

Not a bad idea. I think a user suggested this several years ago if I remember right. It feels maybe a little too magic to me, but it does nicely encapsulate exactly the sugar that's wanted — a synthesized parameter list of initializing formals — and gives users maximum flexibility to use that sugar how they want.